src/west/manifest.py

```diff
@@ -150,1504 +427,1302 @@ def _flags_ok(flags):
-    - ``url``: project fetch URL
-    - ``revision``: revision to fetch from ``url`` when the
-      project is updated
-    - ``path``: relative path to the project within the workspace
-      (i.e. from ``topdir`` if that is set)
-    - ``abspath``: absolute path to the project in the native path name
-      format (or ``None`` if ``topdir`` is)
-    - ``posixpath``: like ``abspath``, but with slashes (``/``) as
-      path separators
-    - ``clone_depth``: clone depth to fetch when first cloning the
-      project, or ``None`` (the revision should not be a SHA
-      if this is used)
-    - ``west_commands``: list of places to find extension commands in
-      the project
-    - ``topdir``: the top level directory of the west workspace
-      the project is part of, or ``None``
-    - ``remote_name``: the name of the remote which should be set up
-      when the project is being cloned (default: 'origin')
-    '''
+        # Add this manifest's projects to the map, then project imports.
+        url_bases = {r['name']: r['url-base'] for r in
+                     manifest.get('remotes', [])}
+        defaults = self._load_defaults(manifest.get('defaults', {}), url_bases)
+        self._load_projects(manifest, url_bases, defaults, ctx)
 
-    def __eq__(self, other):
-        return NotImplemented
+        # The manifest is resolved. Make sure paths are unique.
+        self._check_paths_are_unique(mp, ctx.projects, top_level)
 
-    def __repr__(self):
-        return (f'Project("{self.name}", "{self.url}", '
-                f'revision="{self.revision}", path={repr(self.path)}, '
-                f'clone_depth={self.clone_depth}, '
-                f'west_commands={self.west_commands}, '
-                f'topdir={repr(self.topdir)})')
+        # Save the results.
+        self._projects = list(ctx.projects.values())
+        self._projects.insert(MANIFEST_PROJECT_INDEX, mp)
+        self._projects_by_name: Dict[str, Project] = {'manifest': mp}
+        self._projects_by_name.update(ctx.projects)
+        self._projects_by_cpath: Dict[str, Project] = {}
+        if self.topdir:
+            for i, p in enumerate(self.projects):
+                if i == MANIFEST_PROJECT_INDEX and not p.abspath:
+                    # When from_data() is called without a path hint, mp
+                    # can have a topdir but no path, and thus no abspath.
+                    continue
+                self._projects_by_cpath[util.canon_path(p.abspath)] = p
 
-    def __str__(self):
-        path_repr = repr(self.abspath or self.path)
-        return f'<Project {self.name} ({path_repr}) at {self.revision}>'
+        _logger.debug(f'loaded {loading_what}')
 
-    def __init__(self, name, url, revision=None, path=None,
-                 clone_depth=None, west_commands=None, topdir=None,
-                 remote_name=None):
-        '''Project constructor.
+    def _load_self(self, manifest: Dict[str, Any], path_hint: Optional[str],
```

Comment:
Regarding lists or dictionaries you may also explore using https://docs.python.org/3/library/typing.html#generics (i.e. `Iterable, Mapping...`.

Reply:
The annotations applied to internal data fields are the concrete types returned by the yaml module's load function. I looked for public APIs involving sequence types, and came up with this:

- Manifest.get_projects() and Manifest.projects() return List[Project], but the "list" here is part of the documented API; these don't return a tuple or any other iterable or sequence, but a real list
- Manifest.as_dict() and Manifest.as_frozen_dict() return Dict, which is similarly part of the contract
- Similar comments apply to Project.listdir_at(): the return value should be a list
- Project.git()'s extra_args kwarg is an Iterable already, as is Manifest.get_projects() "project_ids" kwarg

So I think argument types are already permissive in terms of what they accept, but deliberately conservative in terms of what they return. That matches my intent for their behavior. Hopefully that's what you meant.